controller/src/main/java/org/jboss/as/controller/SimpleMapAttributeDefinition.java

```diff
@@ -65,6 +67,37 @@ protected void addOperationParameterValueTypeDescription(ModelNode node, String
         node.get(ModelDescriptionConstants.EXPRESSIONS_ALLOWED).set(new ModelNode(isAllowExpression()));
     }
 
+    @Override
+    public void addCapabilityRequirements(OperationContext context, Resource resource, ModelNode attributeValue) {
+        handleCapabilityRequirements(context, resource, attributeValue, false);
+    }
+
+    @Override
+    public void removeCapabilityRequirements(OperationContext context, Resource resource, ModelNode attributeValue) {
+        handleCapabilityRequirements(context, resource, attributeValue, true);
+    }
+
+    private void handleCapabilityRequirements(OperationContext context, Resource resource, ModelNode attributeValue, boolean remove) {
+        CapabilityReferenceRecorder refRecorder = getReferenceRecorder();
+        if (refRecorder != null && attributeValue.isDefined()) {
+            Set<String> KeyList = attributeValue.keys();
+            String[] attributeValues = new String[KeyList.size()];
```

Comment:
Is that an issue if the `attributeValues` array is not fully filled when `add|removeCapabilityRequirements` is called?

Reply:
I think I borrowed the general pattern from the StringList resource so should be similar.